File: frontend/src/types/patient.ts
export type Gender = 'male' | 'female' | 'other';
export type PatientStatus = 'active' | 'inactive' | 'deceased' | 'transferred';

export interface Caregiver {
  id: string;
  patient_id: string;
  relationship: 'mother' | 'father' | 'guardian' | 'other';
  first_name: string;
  last_name: string;
  contact_info: {
    phone: string;
    alternative_phone?: string;
    email?: string;
    address?: {
      street: string;
      city: string;
      county: string;
      postal_code?: string;
    };
  };
  demographics: {
    age?: number;
    gender?: 'male' | 'female' | 'other';
    education_level?: string;
    occupation?: string;
    language_preference: string;
  };
  engagement: {
    preferred_contact_method: 'sms' | 'whatsapp' | 'voice' | 'email';
    reminder_preferences: {
      days_before: number[];
      time_of_day: string;
    };
    last_contact_date?: string;
    response_rate: number;
  };
  created_at: string;
  updated_at: string;
  status: 'active' | 'inactive';
}

export interface Patient {
  id: string;
  mrn: string; // Medical Record Number
  first_name: string;
  last_name: string;
  date_of_birth: string;
  gender: 'male' | 'female' | 'other';
  contact_info: {
    phone: string;
    alternative_phone?: string;
    email?: string;
    address: {
      street: string;
      city: string;
      county: string;
      postal_code?: string;
      coordinates?: {
        latitude: number;
        longitude: number;
      };
    };
  };
  demographics: {
    ethnicity?: string;
    language_preference: string;
    education_level?: string;
    occupation?: string;
    insurance_provider?: string;
    insurance_number?: string;
  };
  medical_history: {
    allergies: string[];
    chronic_conditions: string[];
    previous_surgeries: string[];
    family_history: string[];
    current_medications: string[];
  };
  biometrics?: {
    fingerprint_id?: string;
    facial_id?: string;
    photo_url?: string;
  };
  caregivers: Caregiver[];
  created_at: string;
  updated_at: string;
  last_visit_date?: string;
  next_visit_date?: string;
  status: 'active' | 'inactive' | 'deceased';
  flags: {
    incomplete_record: boolean;
    high_risk: boolean;
    special_needs: boolean;
    requires_follow_up: boolean;
  };
}

export interface PatientCreate {
  nhif_number?: string;
  biometric_id?: string;
  first_name: string;
  last_name: string;
  date_of_birth: string;
  gender: Gender;
  nationality: string;
  language_preference: string;
  phone_number: string;
  alternative_phone?: string;
  email?: string;
  address: string;
  county: string;
  sub_county: string;
  blood_type?: string;
  allergies?: string[];
  chronic_conditions?: string[];
  medical_history?: Record<string, any>;
  primary_caregiver?: Omit<Caregiver, 'id' | 'registration_date' | 'last_updated' | 'is_active' | 'metadata'>;
}

export interface PatientUpdate {
  nhif_number?: string;
  biometric_id?: string;
  first_name?: string;
  last_name?: string;
  date_of_birth?: string;
  gender?: Gender;
  nationality?: string;
  language_preference?: string;
  phone_number?: string;
  alternative_phone?: string;
  email?: string;
  address?: string;
  county?: string;
  sub_county?: string;
  blood_type?: string;
  allergies?: string[];
  chronic_conditions?: string[];
  medical_history?: Record<string, any>;
  status?: PatientStatus;
  primary_caregiver_id?: string;
}

export interface MedicalRecord {
  id: string;
  patient_id: string;
  record_type: string;
  record_date: string;
  diagnosis?: string;
  treatment?: string;
  notes?: string;
  attachments?: string[];
  created_at: string;
  updated_at: string;
  metadata?: Record<string, any>;
}

export interface Immunization {
  id: string;
  patient_id: string;
  vaccine_name: string;
  scheduled_date: string;
  administered_date?: string;
  batch_number?: string;
  administered_by?: string;
  status: string;
  certificate_url?: string;
  created_at: string;
  updated_at: string;
  metadata?: Record<string, any>;
}

export interface PatientImportResult {
  success: boolean;
  total_records: number;
  imported_records: number;
  failed_records: number;
  errors: {
    row: number;
    error: string;
  }[];
}

export interface PatientSearchParams {
  query: string;
  filters?: {
    status?: Patient['status'];
    age_range?: {
      min: number;
      max: number;
    };
    last_visit_date?: {
      from: string;
      to: string;
    };
    flags?: {
      incomplete_record?: boolean;
      high_risk?: boolean;
      special_needs?: boolean;
      requires_follow_up?: boolean;
    };
  };
  page: number;
  limit: number;
} 

File: backend/models/patient.py
from sqlalchemy import Column, String, Date, Boolean, ForeignKey, JSON, DateTime, Integer, Text, Enum, LargeBinary
from sqlalchemy.orm import relationship
from datetime import date, datetime
import enum
from .base import BaseModel
from ..database import Base

class DigitalConsent(Base):
    __tablename__ = "digital_consents"
    id = Column(Integer, primary_key=True, index=True)
    patient_id = Column(Integer, ForeignKey("patients.id"))
    consent_text = Column(Text, nullable=False)
    consent_given = Column(Boolean, default=False)
    consented_at = Column(DateTime, default=datetime.utcnow)
    consented_by = Column(String(100))  # Name or user ID
    method = Column(String(50))  # e.g., digital, paper, verbal
    metadata = Column(JSON)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    # Relationships
    patient = relationship("Patient", back_populates="digital_consent")

class PatientAuditLog(Base):
    __tablename__ = "patient_audit_logs"
    id = Column(Integer, primary_key=True, index=True)
    patient_id = Column(Integer, ForeignKey("patients.id"))
    action = Column(String(50))  # create, update, delete
    performed_by = Column(String(100))  # user ID or name
    timestamp = Column(DateTime, default=datetime.utcnow)
    changes = Column(JSON)  # Dict of changed fields
    ip_address = Column(String(45))
    user_agent = Column(String(200))
    # Relationships
    patient = relationship("Patient", back_populates="audit_logs")

class BiometricType(str, enum.Enum):
    FINGERPRINT = "fingerprint"
    FACIAL = "facial"
    IRIS = "iris"
    VOICE = "voice"

class BiometricData(Base):
    """Model for storing patient biometric data"""
    __tablename__ = "biometric_data"

    id = Column(Integer, primary_key=True, index=True)
    patient_id = Column(Integer, ForeignKey("patients.id"), nullable=False)
    biometric_type = Column(Enum(BiometricType), nullable=False)
    data = Column(LargeBinary)  # Encrypted biometric data
    nhif_id = Column(String)  # NHIF ID if linked
    capture_date = Column(DateTime, default=datetime.utcnow)
    is_verified = Column(Boolean, default=False)
    verification_date = Column(DateTime)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    patient = relationship("Patient", back_populates="biometric_data")

class PatientPhoto(Base):
    """Model for storing patient photos"""
    __tablename__ = "patient_photos"

    id = Column(Integer, primary_key=True, index=True)
    patient_id = Column(Integer, ForeignKey("patients.id"), nullable=False)
    photo_data = Column(LargeBinary)  # Encrypted photo data
    photo_type = Column(String)  # profile, id_card, etc.
    capture_date = Column(DateTime, default=datetime.utcnow)
    is_approved = Column(Boolean, default=False)
    approval_date = Column(DateTime)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    patient = relationship("Patient", back_populates="photos")

class BulkImport(Base):
    """Model for tracking bulk patient imports"""
    __tablename__ = "bulk_imports"

    id = Column(Integer, primary_key=True, index=True)
    file_name = Column(String, nullable=False)
    file_type = Column(String)  # excel, csv, etc.
    total_records = Column(Integer)
    processed_records = Column(Integer, default=0)
    successful_records = Column(Integer, default=0)
    failed_records = Column(Integer, default=0)
    status = Column(String)  # pending, processing, completed, failed
    error_log = Column(Text)
    created_by = Column(Integer, ForeignKey("users.id"), nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    creator = relationship("User", back_populates="bulk_imports")

class Gender(str, enum.Enum):
    MALE = "male"
    FEMALE = "female"
    OTHER = "other"

class PatientStatus(str, enum.Enum):
    ACTIVE = "active"
    INACTIVE = "inactive"
    DECEASED = "deceased"
    TRANSFERRED = "transferred"

class Patient(Base, BaseModel):
    """Patient model for storing patient information"""
    
    # Basic Information
    first_name = Column(String(100), nullable=False)
    last_name = Column(String(100), nullable=False)
    date_of_birth = Column(Date, nullable=False)
    gender = Column(Enum(Gender))
    
    # Contact Information
    phone_number = Column(String(20))
    email = Column(String(100))
    address = Column(String(200))
    
    # Medical Information
    nhif_number = Column(String(50))
    blood_type = Column(String(5))
    allergies = Column(JSON)
    chronic_conditions = Column(JSON)
    
    # Additional Information
    biometric_id = Column(String(100))  # For fingerprint/facial recognition
    photo_url = Column(String(200))  # URL to patient photo
    notes = Column(String(500))
    
    # Add photo path field
    photo_path = Column(String, nullable=True)
    
    # Status
    is_active = Column(Boolean, default=True)
    last_visit_date = Column(Date)
    
    # Record flagging fields
    is_incomplete = Column(Boolean, default=False)
    is_outdated = Column(Boolean, default=False)
    last_verified_at = Column(DateTime)
    
    # Digital consent
    consent_id = Column(Integer, ForeignKey("digital_consents.id"), nullable=True)
    
    # Relationships
    appointments = relationship("Appointment", back_populates="patient")
    caregivers = relationship("Caregiver", secondary="patient_caregiver", back_populates="patients")
    medical_records = relationship("MedicalRecord", back_populates="patient")
    biometric_records = relationship("BiometricRecord", back_populates="patient")
    follow_ups = relationship("FollowUpSchedule", back_populates="patient")
    notifications = relationship("Notification", back_populates="patient")
    responses = relationship("Response", back_populates="patient")
    
    # Language preferences
    preferred_language = Column(String(10), default="en")  # ISO 639-1 language code
    secondary_languages = Column(JSON)  # List of additional languages
    
    digital_consent = relationship("DigitalConsent", back_populates="patient", uselist=False, foreign_keys=[consent_id])
    audit_logs = relationship("PatientAuditLog", back_populates="patient")
    biometric_data = relationship("BiometricData", back_populates="patient")
    photos = relationship("PatientPhoto", back_populates="patient")
    
    def __repr__(self):
        return f"<Patient {self.first_name} {self.last_name}>"
    
    @property
    def full_name(self):
        return f"{self.first_name} {self.last_name}"
    
    @property
    def age(self):
        today = date.today()
        return today.year - self.date_of_birth.year - (
            (today.month, today.day) < (self.date_of_birth.month, self.date_of_birth.day)
        ) 

# Add relationships to Patient model
from .user import User
User.bulk_imports = relationship("BulkImport", back_populates="creator")

class Caregiver(Base):
    __tablename__ = "caregivers"

    id = Column(String(36), primary_key=True, index=True)
    first_name = Column(String(100))
    last_name = Column(String(100))
    relationship = Column(String(50))
    phone_number = Column(String(20))
    alternative_phone = Column(String(20), nullable=True)
    email = Column(String(100), nullable=True)
    address = Column(String(200))
    county = Column(String(100))
    sub_county = Column(String(100))
    
    # System Fields
    registration_date = Column(DateTime, default=datetime.utcnow)
    last_updated = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    is_active = Column(Boolean, default=True)
    metadata = Column(JSON, nullable=True)
    
    # Relationships
    patients = relationship("Patient", back_populates="primary_caregiver")
    notifications = relationship("Notification", back_populates="caregiver")
    
    def __repr__(self):
        return f"<Caregiver {self.first_name} {self.last_name}>"

class MedicalRecord(Base):
    __tablename__ = "medical_records"

    id = Column(String(36), primary_key=True, index=True)
    patient_id = Column(String(36), ForeignKey("patients.id"))
    record_type = Column(String(50))
    record_date = Column(DateTime)
    diagnosis = Column(String(500), nullable=True)
    treatment = Column(String(500), nullable=True)
    notes = Column(String(1000), nullable=True)
    attachments = Column(JSON, nullable=True)
    
    # System Fields
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    metadata = Column(JSON, nullable=True)
    
    # Relationships
    patient = relationship("Patient", back_populates="medical_records")
    
    def __repr__(self):
        return f"<MedicalRecord {self.record_type} for Patient {self.patient_id}>"

class Immunization(Base):
    __tablename__ = "immunizations"

    id = Column(String(36), primary_key=True, index=True)
    patient_id = Column(String(36), ForeignKey("patients.id"))
    vaccine_name = Column(String(100))
    scheduled_date = Column(DateTime)
    administered_date = Column(DateTime, nullable=True)
    batch_number = Column(String(50), nullable=True)
    administered_by = Column(String(100), nullable=True)
    status = Column(String(20))
    certificate_url = Column(String(200), nullable=True)
    
    # System Fields
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    metadata = Column(JSON, nullable=True)
    
    # Relationships
    patient = relationship("Patient", back_populates="immunizations")
    
    def __repr__(self):
        return f"<Immunization {self.vaccine_name} for Patient {self.patient_id}>" 